crucible-llvm/src/Lang/Crucible/LLVM/MemModel/Type.hs

```diff
@@ -139,8 +137,19 @@ typeEnd a tp = seq a $
     X86_FP80 -> a + 10
     Array 0 _   -> a
     Array n etp -> typeEnd (a + natBytesMul (n-1) (storageTypeSize etp)) etp
-    Struct flds -> typeEnd (a + fieldOffset f) (f^.fieldVal)
-      where f = V.last flds
```

Comment:
Hm, the `Array` case appears to assume that the array has length at least 1:

https://github.com/GaloisInc/crucible/blob/79428012dae4e3c922f75c6f41f418c8aacca665/crucible-llvm/src/Lang/Crucible/LLVM/MemModel/Type.hs#L141

This is admittedly a bit afield of the original scope of the patch, but if the `typeEnd` case for `Struct` is wrong, then it looks wrong for `Array` too.

Reply:
Yeah, it seems like you should be able to make this case underflow using a zero-size array. I'm having a harder time triggering that one. I think `typeEnd` might only be called when you load or store something?

Reply:
> I think `typeEnd` might only be called when you load or store something?

Indeed, and I'm not sure if it's actually possible to load or store a size-0 allocation in a way that invokes `typeEnd`. Nevertheless, I'd sleep better at night knowing that this potential footgun was disarmed. I think we could just add a special case for arrays of size 0 here, much like we now have special cases for empty structs.

Reply:
We both missed that there *is* special handling of 0, right above this:
https://github.com/GaloisInc/crucible/blob/a0ddd876d0d8afba913ccee46dbe6e563f759df3/crucible-llvm/src/Lang/Crucible/LLVM/MemModel/Type.hs#L140

Reply:
Oops, quite right. Thanks for putting my mind at ease.